src/projects/EnsureThat/Enforcers/EnumArg.cs

```diff
@@ -0,0 +1,24 @@
+using System;
+using JetBrains.Annotations;
+
+namespace EnsureThat.Enforcers
+{
+    public sealed class EnumArg
+    {
+        public T IsValidEnum<T>(T value, [InvokerParameterName] string paramName = null, OptsFn optsFn = null) where T : struct, Enum
+        {
+            // TODO: Test performance, not sure if the BCL IsDefined implements this naively or like the fast versino
```

Comment:
Oops, that's on me!

Reply:
?

Reply:
I sent the PR before doing my _own_ TODO :-P Bad form!

Reply:
Soooo problem... A custom implementation of Enum.IsDefined seemed to work a lot faster than the built in version. I wonder 1) what I missed 2) if we should use such a version


Average of Mean | Column   Labels |   |   |  
-- | -- | -- | -- | --
  | LegacyJit |   | RyuJit |  
  | X64 | X86 | X64 |  
Row   Labels | .NET 4.6.2 | .NET 4.6.2 | .NET 4.6.2 | .NET Core 3.1
Custom_IsDefined_No | 7.603 | 10.409 | 16.056 | 13.056
Custom_IsDefined_Yes | 10.396 | 13.8 | 22.134 | 17.762
Custom_IsDefined_Yes_TwiceInARow_WithAnd | 23.248 | 28.557 | 46.46 | 41.889
Enum_IsDefined_No | 253.589 | 300.409 | 474.218 | 242.035
Enum_IsDefined_Yes | 241.672 | 310.138 | 259.76 | 256.527
Enum_IsDefined_Yes_TwiceInARow_WithAnd | 514.641 | 600.283 | 873.107 | 534.571


```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Configs;
using BenchmarkDotNet.Environments;
using BenchmarkDotNet.Jobs;
using BenchmarkDotNet.Running;
using System;

namespace DelMe.IsDefinedTest
{
    using System.Collections.Generic;
    using System.Linq;

    public class Program
    {
        static void Main(string[] args)
        {
            BenchmarkRunner.Run<Program>(
                DefaultConfig.Instance
                    .With(Job.ShortRun.With(Jit.RyuJit).With(CoreRuntime.Core31))
                    .With(Job.ShortRun.With(Jit.RyuJit).With(ClrRuntime.Net462))
                    .With(Job.ShortRun.With(Jit.LegacyJit).With(ClrRuntime.Net462))
                    .With(Job.ShortRun.With(Jit.LegacyJit).With(ClrRuntime.Net462).With(Platform.X86)));
        }

        private Abc yes = Abc.A;
        private Abc no = (Abc) 2;

        [Benchmark]
        public bool Enum_IsDefined_Yes()
        {
            return Enum.IsDefined(typeof(Abc), yes);
        }

        [Benchmark]
        public bool Enum_IsDefined_No()
        {
            return Enum.IsDefined(typeof(Abc), no);
        }

        [Benchmark]
        public bool Enum_IsDefined_Yes_TwiceInARow_WithAnd()
        {
            return Enum.IsDefined(typeof(Abc), yes) && Enum.IsDefined(typeof(Abc), yes);
        }

        [Benchmark]
        public bool Custom_IsDefined_Yes()
        {
            return EnumHelper<Abc>.IsDefined(yes);
        }

        [Benchmark]
        public bool Custom_IsDefined_No()
        {
            return EnumHelper<Abc>.IsDefined(no);
        }

        [Benchmark]
        public bool Custom_IsDefined_Yes_TwiceInARow_WithAnd()
        {
            return EnumHelper<Abc>.IsDefined(yes) && EnumHelper<Abc>.IsDefined(yes);
        }

        private enum Abc
        {
            A
        }
    }

    internal static class EnumHelper<TEnum>
    {
        private static HashSet<TEnum> valid;

        static EnumHelper()
        {
            valid = new HashSet<TEnum>(Enum.GetValues(typeof(TEnum)).Cast<TEnum>());
        }

        public static bool IsDefined(TEnum value)
        {
            return valid.Contains(value);
        }
    }
}
```

Reply:
Ah, evidently they know about the slowness and don't care :)
https://github.com/dotnet/runtime/issues/23179

Reply:
So, @danielwertheim 
- [ ] Custom (trying to optimize) code
- [ ] Custom code, but find it existing in another library
- [ ] Built in slower code

Reply:
As Ensure.That is supposed to be lightweight, I would prefer not to take external dependency.

Your solution was a lot faster, isn't that good enough?

Reply:
I believe so, my concern was mostly if I missed some side cases, but I'll review the implementation details of the BCL method to confirm that isn't the case. There's also the additional memory usage for storing the `valid`, but this should be trivial in almost all applications and in a case it isn't we can handle it better with a user report of an issue anyway (to know if we introduce an overload/change implementation/etc)